cocos/core/platform/view.ts

```diff
@@ -933,13 +934,13 @@ class ContainerStrategy {
         locContainer.style.width = locCanvas.style.width = w + 'px';
         locContainer.style.height = locCanvas.style.height = h + 'px';
         // Setup pixel ratio for retina display
-        let devicePixelRatio = _view._devicePixelRatio = 1;
+        _view._devicePixelRatio = 1;
         if (_view.isRetinaEnabled()) {
-            devicePixelRatio = _view._devicePixelRatio = Math.min(_view._maxPixelRatio, window.devicePixelRatio || 1);
+            _view._devicePixelRatio = Math.min(_view._maxPixelRatio, window.devicePixelRatio || 1);
```

Comment:
This is intend to be freely controlled by user, actually user can set view.maxPixelRatio to a fixed scale, so that the view's pixel ratio will never exceed that.
The benefit is to limit the size of viewport, and pixel fill rate can be controlled.
e.g. we have a device with 3x device pixel ratio, and we can set view.maxPixelRatio to 2, then our viewport will be limited.

Reply:
Good to know. So only use device resolution on native.